Add tracking of users and their accounts

embolalia opened the pull request:
This uses some of the code that @maxpowa wrote for #941, but gives a somewhat more intuitive API. It also paves the way for potentially adding direct support for away-notify and metadata-notify.

Since this being broken could have big security repurcussions (especially since I'd like to be able to make admins/owner pull from account tracking in the future), I'd like a few people to pick this over and see what I've broken before merging it.

embolalia commented:
`bot.channels_` is a sucky name, but idk if anyone is relying on the type of `bot.channels`. Something will have to be done about that before merging.

maxpowa commented:
Privileges could probably be merged into the user object too, so you would do user.privileges[channel] or something along those lines. Would simplify the privileges usage a bit and streamline both paths.

maxpowa commented on sopel/tools/target.py:
```diff
@@ -0,0 +1,64 @@
+    def __eq__(self, other):
+        if not isinstance(other, User):
+            return NotImplemented
+        return self.name == other.name
```
name is not defined for user, same thing on line 28

embolalia commented:
@elad661, @maxpowa: I've addressed the comments, and added some extra features. Do either of you have an objection to replacing the old `bot.channels` with the new dict version? It'll *probably* still work for *most* use cases of the old one, but I'm not sure if it counts as a breaking change. I suppose we could just call it 7.0…

embolalia commented:
I've updated this to get rid of the channels list and replace it with the channels dictionary.

embolalia commented:
@elad661: I kind of figured that's the only kind of use that's common, and it still works (because it's iterating over the keys, which are the same). You're probably right about changing the name of `SopelMemory`.
